### PR TITLE
Client: hive Cancun fixes

### PR DESCRIPTION
**Pull request by jochem-brouwer**
This PR is WIP.

The goal of this PR is to fix most (all?) of failing hive cancun tests. Before this PR: 55 failing tests. After this PR: 30 failing tests

To run:

`./hive --client ethereumjs --sim ethereum/engine --sim.parallelism 8 --sim.limit="engine-cancun"`

(change parallelism and ensure hive is at latest `master`)

Fixes:

- `./hive --client ethereumjs --sim ethereum/engine --sim.parallelism 8 --sim.limit="engine-cancun/Bad Hash on NewPayload"` 104e6e88cdd87bf4c7bbaa8f6ee0955be5fc2649, this fixes 3 tests.
- `Inconsistent Head in ForkchoiceState` in b525804a5e93f64429a1b7ab0e85118defd31c4d, this commit fixes 10 tests. If `recursivelyFindParents` hits genesis block then we should stop finding more parents, we now have a full reorg back to genesis (?)
- `Pre-Merge Fork Number > 0 `: Off by one check in common f2aedcac4ca5cda066733b884a58d0e81ded17ed
- `Unique PayloadID - Withdrawals`: Ensure that different withdrawals result in a different payload ID 14b60a4943e56d09bf4e9975305c4674dc02c6fd

Open failing tests (with notes):

- `Invalid PayloadAttributes: Missing BeaconRoot` -> This test seems wrong, it FCUs without a `parentBlockBeaconRoot` but then expects us to report that blockhash as head hash
- `Invalid Missing Ancestor ReOrg` -> have to do with reorgs, will open a separate PR for this.

### CODE REVIEW

**Comment by jochem-brouwer** [comment]
A client test will fail, this expects us to set finalized/safe to the genesis block if it is not yet set. I am not sure if this is correct. Remainder of this is ready for review, other failing hive tests have to do with reorgs, will open a separate PR for that.

File: packages/common/src/common.ts [hunk 1]
```diff
@@ -384,7 +384,7 @@ export class Common {
       if (mergeIndex >= 0 && td !== undefined && td !== null) {
         if (hfIndex >= mergeIndex && BigInt(hfs[mergeIndex].ttd!) > td) {
           throw Error('Maximum HF determined by total difficulty is lower than the block number HF')
-        } else if (hfIndex < mergeIndex && BigInt(hfs[mergeIndex].ttd!) <= td) {
+        } else if (hfIndex < mergeIndex && BigInt(hfs[mergeIndex].ttd!) < td) {
```
Comment (jochem-brouwer) [inline packages/common/src/common.ts:387]: Ok, this change triggers a Common test. I am pretty sure this is ok, we should check for `<` not `<=`?

**Review by acolytec3 — APPROVED** [review]
LGTM.  I've previously verified that the specified hive tests do indeed pass.

File: packages/common/src/common.ts (hunk 1 above)
Comment (jochem-brouwer) [reply packages/common/src/common.ts:387]: Ah and I forgot to note: this change was triggered by a hive test with thus this exact scenario of hitting the TTD exactly.

**Comment by jochem-brouwer** [comment]
Hmm, I have reran the client tests twice now since I thought this would be a random timeout, but maybe this is not the case (have just retriggered, if it does not pass then I should investigate)

**Comment by jochem-brouwer** [comment]
Ok, tests seems failing into a context which I thought was not relevant, will take a look!

File: packages/client/src/blockchain/chain.ts [hunk 2]
```diff
@@ -294,18 +294,26 @@ export class Chain {
...
+    blocks.finalized = (await this.getCanonicalFinalizedBlock()) ?? null
+    blocks.safe = (await this.getCanonicalSafeBlock()) ?? null
+    blocks.vm = await this.getCanonicalVmHead()
+
+    headers.latest = blocks.latest.header
     // finalized and safe are always blocks since they have to have valid execution
     // before they can be saved in chain
-    headers.finalized = (await this.getCanonicalFinalizedBlock()).header
-    headers.safe = (await this.getCanonicalSafeBlock()).header
+    if (blocks.finalized !== null) {
```
Comment (g11tech) [inline packages/client/src/blockchain/chain.ts:305]:
```suggestion
   headers.finalized = blocks.finalized?.header ?? null
```
should be cleaner and simlarly for `safe`

File: packages/client/src/blockchain/chain.ts [hunk 3]
```diff
@@ -297,13 +297,13 @@ export class Chain {
     headers.latest = await this.getCanonicalHeadHeader()
     // finalized and safe are always blocks since they have to have valid execution
     // before they can be saved in chain
-    headers.finalized = (await this.getCanonicalFinalizedBlock()).header
-    headers.safe = (await this.getCanonicalSafeBlock()).header
+    headers.finalized = blocks.finalized?.header ?? null
+    headers.safe = blocks.safe?.header ?? null
     headers.vm = (await this.getCanonicalVmHead()).header
 
     blocks.latest = await this.getCanonicalHeadBlock()
```
Comment (jochem-brouwer) [inline packages/client/src/blockchain/chain.ts:304]: Ok, so in this file if one first does gets all the blocks and then the headers, then a sync test fails. If one updates this to the original (see 586868ba51d9d3beddd5961fa44d667c4d547d8d ) then it passes. Test in question is `integration/lightclient.spec.ts`

**Comment by holgerd77** [comment]
Reading from the thread it seems this can be merged and will therefore do. If I accidentally merged with some todo still left please eventually just do in a small follow-up PR, but seems convenient to have this in.

**Review by holgerd77 — APPROVED** [review]
LGTM

**Comment by holgerd77** [comment]
(so: pretty cool work with so many Hive fixes! 🤩 🎉)

**Comment by jochem-brouwer** [comment]
Great, thanks :smile: :pray:

File: packages/client/src/blockchain/chain.ts [hunk 4]
```diff
@@ -513,17 +513,17 @@ export class Chain {
   /**
    * Gets the latest block in the canonical chain
    */
-  async getCanonicalSafeBlock(): Promise<Block> {
+  async getCanonicalSafeBlock(): Promise<Block | undefined> {
```
Comment (holgerd77) [inline packages/client/src/blockchain/chain.ts:516]: Just writing releases notes and stumbled upon this: so this is a (type) breaking change.

I first tended to just leave and hope not too many people use this and the method below yet but now when thinking about it twice I would prefer of going (a small) extra round here and introduce two new methods `getCanonicalSafeBlockSafe()` and `getCanonicalFinalizedBlockSafe()` and keep the behavior of the current methods as is, also considering that both methods are just used once in client and can be easily updated.

Comment (jochem-brouwer) [reply packages/client/src/blockchain/chain.ts:516]: Ah wait, I just stumbled upon this again, but this is into our client package (and not in our blockchain package). So I think we can leave this as-is? It is very unlikely someone uses our client as an actual package dependency?

Comment (holgerd77) [reply packages/client/src/blockchain/chain.ts:516]: Ah, you are right. I only saw `blockchain/*`. 🤦‍♂️

Sorry for the non-issue.

**Pull request merged**